### PR TITLE
test(style): Limit allowed attributes on <a> elements

### LINKED ISSUE #4033: Lint for unnnecessary attributes in markup (e.g. inside notes)
As per https://github.com/mdn/browser-compat-data/pull/3981#issuecomment-488245877

We ought to update the linter to flag unneccessary attributes (like title attributes in notes); they aren't useful and mean more of a buden if we try to localize BCD in the future.

### PR DESCRIPTION
**Pull request by ExE-Boss**
Resolves #4033

### CODE REVIEW

File: api/BatteryManager.json [hunk 1]
```diff
@@ -150,7 +150,7 @@
...
+                "notes": "Disabled by default in Firefox 10.0, but can be enabled setting the preference <code>dom.battery.enabled</code> to <code>true</code>. Starting with Firefox 11.0, <code>mozBattery</code> is enabled by default. The Battery API is currently supported on Android, Windows, and Linux with UPower installed. Support for MacOS is available starting with Gecko 18.0 (Firefox 18.0 / Thunderbird 18.0 / SeaMonkey 2.15). Firefox also provide support for the deprecated <a href='https://developer.mozilla.org/docs/Web/API/Navigator/battery'><code>navigator.battery</code></a>."
```
Comment (ExE-Boss) [inline api/BatteryManager.json]: This `title` attribute here was actually unclosed:
```html
Firefox also provide support for the deprecated <a
	href='https://developer.mozilla.org/docs/Web/API/Navigator/battery'
	title='The battery read-only property returns a BatteryManager provides information about
the system's battery charge level; provides also some new events you can handle to monitor
the battery status. This implements the Battery Status API; see that documentation for additional
details, a guide to using the API, and sample code.'>
	<code>navigator.battery</code>
</a>
```

File: api/BatteryManager.json [hunk 2]
```diff
@@ -35,7 +35,7 @@
...
+              "notes": "Disabled by default in Firefox 10, but can be enabled setting the preference <code>dom.battery.enabled</code> to <code>true</code>. Starting with Firefox 11, <code>mozBattery</code> is enabled by default. The Battery API is currently supported on Android, Windows, and Linux with UPower installed. Support for MacOS is available starting with Firefox 18. Firefox also provide support for the deprecated <a href='https://developer.mozilla.org/docs/Web/API/Navigator/battery'><code>navigator.battery</code></a>."
```
Comment (ExE-Boss) [inline api/BatteryManager.json]: The `title` attribute here was actually unclosed:
```html
Firefox also provide support for the deprecated <a
	href='https://developer.mozilla.org/docs/Web/API/Navigator/battery'
	title='The battery read-only property returns a BatteryManager provides
information about the system's battery charge level; provides also some
new events you can handle to monitor the battery status.
This implements the Battery Status API; see that documentation
for additional details, a guide to using the API, and sample code.'>
	<code>navigator.battery</code>
</a>
```

File: api/CharacterData.json [hunk 3]
```diff
@@ -65,7 +65,7 @@
             },
             "firefox": {
               "version_added": "25",
-              "notes": "Two properties, <code>nextElementSibling</code> and <code>previousElementSibling</code>, have been moved to the <a href='https://developer.mozilla.org/docs/Web/API/NonDocumentTypeChildNode><code>NonDocumentTypeChildNode</code></a> interface, also implemented by <code>CharacterData</code>."
+              "notes": "Two properties, <code>nextElementSibling</code> and <code>previousElementSibling</code>, have been moved to the <a href='https://developer.mozilla.org/docs/Web/API/NonDocumentTypeChildNode'><code>NonDocumentTypeChildNode</code></a> interface, also implemented by <code>CharacterData</code>."
```
Comment (ExE-Boss) [inline api/CharacterData.json]: This `href` attribute had an unclosed `'`, which broke the compatibility table: https://developer.mozilla.org/docs/Web/API/CharacterData#Browser_compatibility

File: css/properties/transition.json [hunk 4]
```diff
@@ -47,7 +47,7 @@
...
-                  "Before Firefox 57, the <a href='https://developer.mozilla.org/docs/Web/CSS/background-position'><code>background-position</code></a> property can't be transitioned between two values containing different numbers of <a href='https://developer.mozilla.org/docs/Web/CSS/position_value' t><code>&lt;position&gt;</code></a> values, for example <code>background-position: 10px 10px;</code> and <code>background-position: 20px 20px, 30px 30px;</code> (see <a href='https://bugzil.la/1390446'>bug 1390446</a>)."
+                  "Before Firefox 57, the <a href='https://developer.mozilla.org/docs/Web/CSS/background-position'><code>background-position</code></a> property can't be transitioned between two values containing different numbers of <a href='https://developer.mozilla.org/docs/Web/CSS/position_value'><code>&lt;position&gt;</code></a> values, for example <code>background-position: 10px 10px;</code> and <code>background-position: 20px 20px, 30px 30px;</code> (see <a href='https://bugzil.la/1390446'>bug 1390446</a>)."
```
Comment (ExE-Boss) [inline css/properties/transition.json]: These `<a>` elements had an extra empty `t` attribute for whatever reason.

File: test/test-style.js [hunk 5]
```diff
@@ -44,6 +47,48 @@ function escapeInvisibles(str) {
...
 }
 
+/**
+ * Gets the row and column matching the index in a string.
+ *
+ * @param {string} str
+ * @param {number} index
+ * @return {[number, number] | [null, null]}
+ */
+function indexToPos(str, index) {
```
Comment (ExE-Boss) [inline test/test-style.js]: I imported this helper function from #3913.

**Comment by Elchi3** [comment]
So first of all, thanks for your quick PR on this topic, @ExE-Boss! I think you uncovered some problems that are definitely worth fixing.

In here, you're using some real advanced regex magic and I wonder if this is a good approach or whether we wouldn't be better off by using a HTML validation/sanitization library or so, especially since this isn't the first time this comes up: there are several problems due to us not doing proper validation of free form fields like the `notes` property. However, I feel that this PR is a bit like duct tape so only a subset of problems are addressed and we need more duct tape to address more issues. So, I want to take a step back and look at a few issues together so we can decide how we could solve them:

- notes allow HTML, but we don't validate if it is well-formed HTML (https://github.com/mdn/browser-compat-data/issues/3985)
- notes allow all kinds of HTML, but we only want to allow `<a>` and `<code>` (maybe more) (https://github.com/mdn/browser-compat-data/issues/2799)
- notes allow <a> links, but we only want them to have `href` (#4033 and this PR)

I'm not sure if all this is still best kept in "test-style" or if we need larger weapons to do this properly (maybe "test-html" and the use of a library instead of our own complicated regexes).

What do you think about this topic?

**Comment by ddbeck** [comment]
> In here, you're using some real advanced regex magic and I wonder if this is a good approach or whether we wouldn't be better off by using a HTML validation/sanitization library

FWIW, I used jsdom to check this sort of thing in short-descriptions https://github.com/mdn/short-descriptions/blob/master/test/content-rules.js#L21 It's another option worth considering.

**Comment by vinyldarkscratch** [comment]
Just to throw out additional information, I found [sanitize-html](https://www.npmjs.com/package/sanitize-html#node-recommended) which seems to allow the limitation of HTML elements and attributes on those elements.  I would probably look at this less as a dependency and more as an example of what we're looking for as it performs the fixes and outputs clean HTML (rather than letting us know what the problems are), but we could totally do an `original !== clean` check or something.

**Comment by ExE-Boss** [comment]
Closing in favour of #4259.

**Pull request closed without merging**